corehq/apps/app_manager/tasks.py

```diff
@@ -23,26 +22,40 @@ def create_user_cases(domain_name):
 
 
 @serial_task('{app_id}-{version}', max_retries=0, timeout=60 * 60)
-def make_async_build_v2(app_id, domain, version, allow_prune=True, comment=None):
+def make_async_build_v2(app_id, domain, version, username, allow_prune=True, comment=None):
     app = get_app(domain, app_id)
-    return make_async_build(app, allow_prune=allow_prune, comment=comment)
-
-
-@serial_task('{app._id}-{app.version}', max_retries=0, timeout=60 * 60)
-def make_async_build(app, allow_prune=True, comment=None):
-    previous_version = app.get_previous_version()
-    if previous_version and previous_version.version == app.version:
-        return
     errors = app.validate_app()
     if not errors:
         if allow_prune and comment is None:
             comment = _('Auto-generated by a phone update. Will expire after next build if not marked released.')
+        if username:
+            comment = (comment + ' ') or ''
+            comment += _('Generated by user {}.').format(username)
```

Comment:
The comment generation logic could be simplified signficantly if you push it into `make_async_build` and `make-build`:

```python
def make_async_build(...):
    comment = _('Auto-generated by a phone update. Will expire after next build '
                'if not marked released. Generated by user {}.').format(username)

def make_build(...):
    # comment provided by caller, doesn't need username
```

Actually, `make_build` is only used in one place, to generate a build from a template app.  That means it probably doesn't need the `_should_make_build` check.  It also doesn't need `is_auto_generated` set, or for the app to be saved.  All it really needs is for the app to be validated and built.  I think if you fully split up these two use-cases a lot of this code will be unnecessary.

Reply:
> That means it probably doesn't need the _should_make_build check. It also doesn't need is_auto_generated set, or for the app to be saved.

I'm wary of changing the logic much based only on where these functions are currently called - calling code shouldn't influence a function's internal logic that much. As a reader, I'd expect `make_build` and a `make_async_build` to behave the same except that one is synchronous and one is not. I think it's a different story if `make_build` is named something more distinct, like `make_build_from_template`.


Tangentially, I feel like we talk about naming so much more since people started reading Clean Code. Hooray for book club.

Reply:
Oh, totally agree - this presents itself as a generic function, but that's not how it's used, with a lot of logic specific to one caller.  I thought I mentioned something about renaming for the specific use-cases, but it looks like I didn't.

By struggling too hard to be generic, it ends up needing to support workflows that don't currently happen in practice (like `username and not comment`).  Plus, things like `Auto-generated by a phone update` are specific to one of the workflows and ideally wouldn't be in a generic function.

Reply:
> It also doesn't need is_auto_generated set

that would change the logic and i am not sure of doing that and would go out of scope of this PR.

Reply:
Comment moved out in https://github.com/dimagi/commcare-hq/pull/24860/commits/76a7fd31035703f9c2cbdaa521c23109e861cca4